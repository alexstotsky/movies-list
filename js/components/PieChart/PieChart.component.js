import React from 'react'
import {
  StyleSheet,
  Text,
  View,
  ART,
  TouchableWithoutFeedback,
} from 'react-native'
import PropTypes from 'prop-types'

import * as scale from 'd3-scale'
import * as shape from 'd3-shape'

import AnimatedShape from '../AnimatedShape'

import COLORS from '../../constants/colors'
import { getRandomColor } from '../../utils/color'

const {
  Surface,
  Group,
} = ART

const d3 = {
  scale,
  shape,
}

const margin = 10

class PieChart extends React.Component {
  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
      number: PropTypes.number,
      name: PropTypes.string,
    })).isRequired,
    pieWidth: PropTypes.number.isRequired,
    pieHeight: PropTypes.number.isRequired,
    innerRadius: PropTypes.number.isRequired,
    selectedPadding: PropTypes.number.isRequired,
  }

  constructor(props) {
    super(props)
    this.state = {
      highlightedIndex: 0,
    }
    this.colors = props.data.map((item, index) => COLORS.chart[index] || getRandomColor())
  }

  _value = item => item.number

  _label = item => item.name

  _color = index => this.colors[index]

  _createPieChart = (index) => {
    const {
      data, pieWidth, innerRadius, selectedPadding,
    } = this.props
    const { highlightedIndex } = this.state
    const arcs = d3.shape.pie()
      .value(this._value)(data)

    const hightlightedArc = d3.shape.arc()
      .outerRadius(pieWidth / 2)
      .padAngle(0.05)
      .innerRadius(innerRadius)

    const arc = d3.shape.arc()
      .outerRadius(pieWidth / 2 - selectedPadding)
      .padAngle(0.05)
      .innerRadius(innerRadius)

    const arcData = arcs[index]
    const path = (highlightedIndex === index) ? hightlightedArc(arcData) : arc(arcData)

    return {
      path,
      color: this._color(index),
    }
  }

  _onPieItemSelected = (index) => {
    const { onItemSelected = () => {} } = this.props
    this.setState({
      highlightedIndex: index,
    })
    onItemSelected(index)
  }

  render() {
    const {
      data, pieWidth, pieHeight,
    } = this.props
    const { highlightedIndex } = this.state
    const x = (pieWidth + margin) / 2
    const y = (pieWidth + margin) / 2
    return (
      <View style={styles.container}>
        <View style={{ width: pieWidth + margin, height: pieHeight + margin }}>
          <Surface
            width={pieWidth + margin}
            height={pieHeight + margin}
          >
            <Group x={x} y={y}>
              {
                data.map((item, index) => (
                  <AnimatedShape
                    key={`pie_chart_${index}`}
                    color={this._color(index)}
                    d={() => this._createPieChart(index)}
                  />
                ))
              }
            </Group>
          </Surface>
        </View>
        <View style={styles.legend}>
          {
            data.map((item, index) => {
              const fontWeight = highlightedIndex === index ? 'bold' : 'normal'
              return (
                <TouchableWithoutFeedback key={index} onPress={() => this._onPieItemSelected(index)}>
                  <View>
                    <Text
                      style={[
                        styles.label,
                        {
                          color: this._color(index),
                          fontWeight,
                        },
                      ]}
                    >
                      {`${this._label(item)}: ${this._value(item)}%`}
                    </Text>
                  </View>
                </TouchableWithoutFeedback>
              )
            })
          }
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  legend: {
    flex: 1,
    paddingLeft: 10,
  },
  label: {
    marginTop: 5,
  },
})

export default PieChart
