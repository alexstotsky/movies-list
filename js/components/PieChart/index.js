import PieChart from './PieChart.component'

export default PieChart
