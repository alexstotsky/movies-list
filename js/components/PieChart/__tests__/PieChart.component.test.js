import React from 'react'
import { shallow } from 'enzyme'

import PieChart from '../PieChart.component'

describe('Testing PieChart component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <PieChart />,
    )
    expect(wrapper).toMatchSnapshot()
    wrapper.setProps({
      data: [
        {
          number: 15,
          name: '15',
        },
      ],
      selectedPadding: 10,
      innerRadius: 20,
      pieHeight: 15,
      pieWidth: 15,
    })
    expect(wrapper).toMatchSnapshot()
  })
})
