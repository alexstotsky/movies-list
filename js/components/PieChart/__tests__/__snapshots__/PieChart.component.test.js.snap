// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PieChart component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "height": 50,
        "width": 50,
      }
    }
  >
    <Surface
      height={50}
      width={50}
    >
      <Group
        x={25}
        y={25}
      />
    </Surface>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingLeft": 10,
      }
    }
  />
</View>
`;

exports[`Testing PieChart component renders correctly 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "height": 25,
        "width": 25,
      }
    }
  >
    <Surface
      height={25}
      width={25}
    >
      <Group
        x={12.5}
        y={12.5}
      >
        <AnimatedShape
          d={[Function]}
          key="pie_chart_0"
        />
      </Group>
    </Surface>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingLeft": 10,
      }
    }
  >
    <TouchableWithoutFeedback
      key="0"
      onPress={[Function]}
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "marginTop": 5,
              },
              Object {
                "color": undefined,
                "fontWeight": "bold",
              },
            ]
          }
        >
          15: 15%
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;
